src/libraries/Common/src/Interop/Windows/SspiCli/SecuritySafeHandles.cs

```diff
@@ -323,45 +324,54 @@ public new IntPtr DangerousGetHandle()
     //
     // This is a class holding a Credential handle reference, used for static handles cache
     //
-#if DEBUG
-    internal sealed class SafeCredentialReference : DebugCriticalHandleMinusOneIsInvalid
-    {
-#else
-    internal sealed class SafeCredentialReference : CriticalHandleMinusOneIsInvalid
+    internal sealed class SafeCredentialReference : CriticalFinalizerObject, IDisposable
     {
-#endif
-
         //
         // Static cache will return the target handle if found the reference in the table.
         //
         internal SafeFreeCredentials Target;
 
         internal static SafeCredentialReference? CreateReference(SafeFreeCredentials target)
         {
-            SafeCredentialReference result = new SafeCredentialReference(target);
-            if (result.IsInvalid)
+            if (target.IsInvalid || target.IsClosed)
             {
                 return null;
             }
 
+            SafeCredentialReference result = new SafeCredentialReference(target);
+
             return result;
         }
-        private SafeCredentialReference(SafeFreeCredentials target) : base()
+        private SafeCredentialReference(SafeFreeCredentials target)
         {
             // Bumps up the refcount on Target to signify that target handle is statically cached so
             // its dispose should be postponed
             bool ignore = false;
             target.DangerousAddRef(ref ignore);
             Target = target;
-            SetHandle(new IntPtr(0));   // make this handle valid
         }
 
-        protected override bool ReleaseHandle()
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void Dispose(bool disposing)
         {
             SafeFreeCredentials target = Target;
             target?.DangerousRelease();
             Target = null!;
-            return true;
+        }
+
+        ~SafeCredentialReference()
+        {
+            Dispose(false);
+
+#if DEBUG
+            Debug.Fail("Finalizer should never be called. The SafeCredentialReference should be always disposed manually.");
+            throw NotImplemented.ByDesign;
+#endif
```

Comment:
Why is this valid?  There's no way user code could drop one of these on the floor and result in this being invoked?

Even if the answer is "no", this assert could fire / exception thrown if the SafeFreeCredentials instance passed to the constructor is concurrently closed after target.IsClosed check in CreateReference and before the DangerousAddRef call in the ctor.  If that happens, the DangerousAddRef will throw an exception, but by that point, the SafeCredentialReference instance will have already been allocated and its finalizer will eventually run.

Seems like this whole if DEBUG / endif section should be deleted.

Reply:
I have removed ifdefs